### PR TITLE
add ExcludePaths, to make the SignalR endpoint take effect.

### PR DESCRIPTION
**Pull request by chen1tian**
# Support for SignalR
if you have a SigalR EndPoint，like：
```csharp
app.UseEndpoints(endpoints =>
{
    endpoints.MapControllers();
    endpoints.MapHub<NoticeHub>("/notice");
});
```

then you can use ExcludePaths exclude it to make the SignalR endpoint take effect.
```csharp
app.UseCustomAutoWrapper(new AutoWrapper.AutoWrapperOptions
{      
    ExcludePaths = new string[] { "/notice" }
});
```

### CODE REVIEW

File: src/AutoWrapper/Base/WrapperBase.cs [hunk 1]
```diff
@@ -50,6 +52,9 @@ public virtual async Task InvokeAsyncBase(HttpContext context, AutoWrapperMember
 
                     if (context.Response.HasStarted) { LogResponseHasStartedError(); return; }
 
+                    var url = Microsoft.AspNetCore.Http.Extensions.UriHelper.BuildRelative(context.Request.Path);
```
Comment (hueifeng) [inline src/AutoWrapper/Base/WrapperBase.cs:55]: There seems to be no call to it, so consider cleaning it up.

File: README.zh-cn.md [hunk 2]
```diff
@@ -21,6 +21,7 @@
   * 添加对问题详细信息的异常格式的支持
   * 添加对忽略不需要使用`[AutoWrapIgnore]`过滤器属性包装的操作方法的支持。
   * V3.x启用了对`netcoreapp2.1`和`netcoreapp2.2` .NET Core框架的向后兼容性支持
+  * 增加排除的路径，依赖于此，增加了对SignalR的支持。
```
Comment (hueifeng) [inline README.zh-cn.md:24]: Do you think it is possible to change the description text to "支持对指定路径集的忽略，同时适用于SignalR."?  It seems that the features of the English document are not written.

**Review by hueifeng — COMMENTED** [review]
Thank you for your PR, it is great.

File: src/AutoWrapper/Base/WrapperBase.cs (hunk 1 above)
Comment (chen1tian) [reply src/AutoWrapper/Base/WrapperBase.cs]: ok, cleaned up.

File: README.zh-cn.md (hunk 2 above)
Comment (chen1tian) [reply README.zh-cn.md]: yes, added now.

**Comment by proudmonkey** [comment]
@chen1tian Thanks for the PR. This is a great addition to AutoWrapper feature! I will merge this if you are confident with it. In the upcoming .NET 5 conversion, I think we should merge the `/swagger` path to this new feature as the default exclude path. What do you think?

File: src/AutoWrapper/AutoWrapperMembers.cs [hunk 3]
```diff
@@ -214,10 +216,33 @@ public bool IsSwagger(HttpContext context, string swaggerPath)
...
+            {
+                switch (x.ExcludeMode)
+                {
+                    default:
+                    case ExcludeMode.Strict:
+                        return context.Request.Path.Value == x.Path;
+                    case ExcludeMode.StartWith:
+                        return context.Request.Path.StartsWithSegments(new PathString(x.Path));
+                    case ExcludeMode.Regex:
+                        Regex regExclue = new Regex(x.Path);
```
Comment (proudmonkey) [inline src/AutoWrapper/AutoWrapperMembers.cs:236]: Typo here. Should be `regExclude`

File: src/AutoWrapper/AutoWrapperExcludePaths.cs [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+namespace AutoWrapper
+{
+    public enum ExcludeMode
+    {
+        Strict = 1,
+        StartWith = 2,
+        Regex = 3
+    }
+
+    public class AutoWrapperExcludePaths
```
Comment (proudmonkey) [inline src/AutoWrapper/AutoWrapperExcludePaths.cs:14]: Can we make this class in singular form?

File: src/AutoWrapper/AutoWrapperMembers.cs (hunk 3 above)
Comment (chen1tian) [reply src/AutoWrapper/AutoWrapperMembers.cs:236]: yes，i will fix it.

File: src/AutoWrapper/AutoWrapperExcludePaths.cs (hunk 4 above)
Comment (chen1tian) [reply src/AutoWrapper/AutoWrapperExcludePaths.cs:14]: sure, I will fix it.

**Comment by proudmonkey** [comment]
Thank you, @chen1tian - Merged and added you to the contributor list :)

**Pull request merged**